docs/testing/writing-unit-tests.md

```diff
@@ -0,0 +1,456 @@
+# Writing Narrative front end unit tests
+
+***Table of Contents***
+  - [Running tests](#running-tests)
+  - [Unit test frameworks](#unit-test-frameworks)
+    - [Jasmine](#jasmine)
+    - [Karma](#karma)
+    - [Configuration](#configuration)
+  - [Writing test specs](#writing-test-specs)
+    - [Getting started](#getting-started)
+    - [Set up and tear down](#set-up-and-tear-down)
+    - [Asynchrony](#asynchrony)
+    - [Mocking and Spies](#mocking-and-spies)
+  - [Narrative Mocks Library](#narrative-mocks-library)
+    - [Loading and using the library](#loading-and-using-the-library)
+  - [KBase Narrative Unit Test Best Practices](#kbase-narrative-unit-test-best-practices)
+
+This document details how to write and run front end unit tests for the KBase Narrative. For information about integration tests look [here](integration-testing.md). For general test running, look [here](./testing.md).
+
+This document assumes you have a working Narrative installation.
+
+## Running tests
+The simple way to run tests is by just running `make test-frontend-unit`. This handles marshalling everything, running the tests, and shutting it all back down again.
+
+Doing it manually requires two phases.
+
+In the first phase, a running Narrative instance must be set up on port 32323. The easiest way to do this is with the command `kbase-narrative --no-browser --port=32323`. This provides a working Narrative and makes the Jupyter Notebook code available on that port, which is configured into karma for its proxies.
+
+The second phase is running the tests themselves. Just run `npm run test`, which will start Karma and run the test suite.
+
+
+## Unit test frameworks
+Narrative front end unit testing is done using the [Jasmine](https://jasmine.github.io/) test framework, and run using the [Karma](https://karma-runner.github.io/). These are configured to load and proxy the Jupyter notebook frontend as well as the KBase code extensions that make the Narrative work.
+
+### Jasmine
+
+Jasmine is a popular Javascript unit test framework. It makes use of common JS testing paradigms and can handle testing asynchronous code.
+
+### Karma
+
+Karma is a test runner that was primarily made to support the Angular framework, but works well outside of that. At it's core, Karma starts a small HTTP server to serve the test module and provide proxy access to both KBase and Jupyter Javascript code modules.
+
+### Configuration
+
+Karma configuration is done under `test/unit/karma.conf.js`. As used for Narrative testing, this sets up the following:
+* A set of browsers to test on
+* KBase modules files to preprocess for coverage reporting
```

Comment:
either "KBase module files", "KBase modules" or "KBase files"

Reply:
I went with "KBase module files" to emphasize that this should be a list of code files that get loaded by the server.